harvester/node-disk-manager
# Misc NDM Improvement

**Problem:**
1. Now we find out that NDM will remove the hot-plug device from the `blockdevices` that would cause the data to be lost when it add back.
2. Also, we allow the user to use the repair tool instead of the force format again.
3. Some improvement

**Solution:**
1. We changed the state to inactive for the hotplugged device instead of deleting it from `blockdevices` CR.
2. Add new field `corrupted`. For the `corrupted` device, we could choose the force format or try to use e2fsck to repair.
3. Log improvement
4. reduce the redundant CR update
5. remove the polling mechanism (But still keep two handling function now)

**Related Issue:**
https://github.com/harvester/harvester/issues/3218
https://github.com/harvester/harvester/issues/3412
https://github.com/harvester/harvester/issues/4016

**Test plan:**
case 1:
1. add a single disk
2. create external storage with the above disk
3. remove this disk
4. check the `blockdevices` status is inactive
5. add this disk back
6. check that the `blockdevices` status is active and should work again.

case 2:
**NOTE1: Please update the blockdevices CRD (on this PR) before we do this test**
**NOTE2: Please ensure you use the SCSI disk with wwn or the filesystem UUID will be erased when we try to break the filesystem.**
1. add a single disk
2. create external storage with the above disk
3. remove this storage
4. make this filesystem corrupted (you could use the following command `dd if=/dev/zero of=/dev/sda bs=512k count=1 oflag=sync`
5. try to create external storage with this disk
6. you should see the mount error and check the `Status.Filesystem.Corrupted` is true
7. Check on the UI that you could force format again.
8. Add it back with force format, and it should work again.
